### PR TITLE
Add Pylons and Forestry to UK Base world

### PR DESCRIPTION
**Pull request by Ushcode**
Data sources, config files and relevant documentation have been added to the data directory of the uk-base-world viz to add the power network assets and the forestry data.

The script that automatically copies the data to the right place and runs the manager and uploader was also updated

### CODE REVIEW

**Comment by Ushcode** [comment]
I'm not sure if things should be getting committed to the `Agents/FeatureInfoAgent/queries/` folder so we'll block them if not

**Comment by sm453** [comment]
> I'm not sure if things should be getting committed to the `Agents/FeatureInfoAgent/queries/` folder so we'll block them if not

As a rule, use-case-specific things should not be in generic folders like `Agents/FeatureInfoAgent/queries/`.

**Comment by mdhillman** [comment]
For the time being, I am going to push this PR back to the draft status as we need to review some of the data licensing before taking any further action.

File: web/twa-vis-framework/example-mapbox-vis/webspace/data/singapore/rail-lines.geojson [hunk 1]
```diff
@@ -1 +1 @@
...83268, 1.3021], [103.83306, 1.30135], [103.83349, 1.30092], [103.8342, 1.30061], [103.83467, 1.30048], [103.83524, 1.30032], [103.83599, 1.30042], [103.83605, 1.30044], [103.83669, 1.30063], [103.83715, 1.3006], [103.83851, 1.30034], [103.84026, 1.29994], [103.84285, 1.29933], [103.84502, 1.29878], [103.84604, 1.29851], [103.85065, 1.29552], [103.8512, 1.29493], [103.85179, 1.29401], [103.85199, 1.29372], [103.85235, 1.29337], [103.85262, 1.29309], [103.85318, 1.29248], [103.85355, 1.29177], [103.85365, 1.29126], [103.85357, 1.29082], [103.85257, 1.28815], [103.85193, 1.28668], [103.85171, 1.28606], [103.85161, 1.28544], [103.85155, 1.28465], [103.85154, 1.28441], [103.85146, 1.28297], [103.85149, 1.28278], [103.85152, 1.28266], [103.85176, 1.28233], [103.85208, 1.2818], [103.85262, 1.28086], [103.85318, 1.27995], [103.85375, 1.27899], [103.85404, 1.27846], [103.85418, 1.27813], [103.8543, 1.27768], [103.85475, 1.27517], [103.8549, 1.27468], [103.85506, 1.27431], [103.85523, 1.27399], [103.85558, 1.27348], [103.8558, 1.27323], [103.85818, 1.27102], [103.85897, 1.27071], [103.85963, 1.27062], [103.86038, 1.27058], [103.86104, 1.27071], [103.86141, 1.27087], [103.86272, 1.27175]]}}]}
```
Comment (mdhillman) [inline web/twa-vis-framework/example-mapbox-vis/webspace/data/singapore/rail-lines.geojson:1]: I can't see a reason for the change in this file, please list reason or revert.

File: web/uk-base-world/scripts/start.sh [hunk 2]
```diff
@@ -43,13 +43,16 @@ then
...
-    FIA_DIR="$ROOT/Agents/FeatureInfoAgent/queries"
-    cp "./inputs/config/fia-config.json" "$FIA_DIR/"
-    cp "./inputs/config/dukes_query.sparql" "$FIA_DIR/"
+    # Copy in the FIA config and query ( removed this bit )
+    # FIA_DIR="$ROOT/Agents/FeatureInfoAgent/queries"
+    # cp "./inputs/config/fia-config.json" "$FIA_DIR/"
+    # cp "./inputs/config/dukes_query.sparql" "$FIA_DIR/"
+    # cp "./inputs/config/power_line_query.sparql" "$FIA_DIR/"
+    # cp "./inputs/config/substation_query.sparql" "$FIA_DIR/"
+    # cp "./inputs/config/woodland_query.sparql" "$FIA_DIR/"
```
Comment (mdhillman) [inline web/uk-base-world/scripts/start.sh:55]: Can safely remove these commented out lines.

File: web/uk-base-world/scripts/start.sh [hunk 3]
```diff
@@ -98,30 +105,35 @@ then
 
     # Copy in the stack uploader config(s)
     cp "./inputs/config/dukes_2023.json" "$UPLOAD_CONFIG/"
+    cp -r ./inputs/config/pylons-and-veg/* "$UPLOAD_CONFIG/"
 
     # Copy in the data for upload
-    UPLOAD_DATA="$ROOT/Deploy/stacks/dynamic/stack-data-uploader/inputs/data"
+    #UPLOAD_DATA="$ROOT/Deploy/stacks/dynamic/stack-data-uploader/inputs/data"
```
Comment (mdhillman) [inline web/uk-base-world/scripts/start.sh:111]: Commenting out this line breaks the subsequent copy calls.

File: web/uk-base-world/README.md [hunk 4]
```diff
@@ -71,12 +82,12 @@ To run the script and bring up a local instance of the UK Base World visualisati
    - Add your API key to a file at `./visualisation/mapbox_api_key`
 4. Add the data files:
    - Add the processed DUKES 2023 CSV to the `./inputs/data/uk_base_world/dukes_2023` directory.
-   - Add the OntoEIP OWL and TTL files to the `./inputs/data/ontoeip` directory (note that these need to be in a flat structure, no subdirectories).
+   - Add the OntoEIP OWL and TTL files to the `./inputs/data/uk_base_world/ontoeip` directory (note that these need to be in a flat structure, no subdirectories).
```
Comment (mdhillman) [inline web/uk-base-world/README.md:85]: Appreciate that you've added READMEs within the specific data folders, but I doubt people will know to look for them. Please add an entry to this list for each new data set that needs copying in.

File: web/uk-base-world/inputs/data/national_grid/vector/cables/README.md [hunk 5]
```diff
@@ -0,0 +1 @@
+Add national grid underground cables ShapeFile files here. Download as .zip from https://connecteddata.nationalgrid.co.uk/dataset/?groups=system-and-network&res_format=ZIP.
```
Comment (mdhillman) [inline web/uk-base-world/inputs/data/national_grid/vector/cables/README.md:1]: National Grid website lists multiple "cables" data sets, please list which one is required (or state that any will do).

Same goes for other READMEs in the data/national_grid directory.

File: web/uk-base-world/scripts/start.sh [hunk 6]
```diff
@@ -98,30 +105,35 @@ then
...
+    echo "Copying data..." 
     cp -r "./inputs/data/." "$UPLOAD_DATA/" 
 
     # Run the stack manager to start a new stack
-    cd "$ROOT/Deploy/stacks/dynamic/stack-manager"
+    cd "$ROOT/Deploy/stacks/dynamic/stack-manager" || exit
     echo "Running the stack start up script..."
-    ./stack.sh start UKBASEWORLD 38383
+    ./stack.sh start UKBASEWORLD "$HOST"
```
Comment (mdhillman) [inline web/uk-base-world/scripts/start.sh:118]: Missing port 38383

Comment (mdhillman) [reply web/uk-base-world/scripts/start.sh:118]: $HOST also not needed here.

File: web/uk-base-world/inputs/config/pylons-and-veg/national-grid.json [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+            "type": "tabular",
+            "skip": false,
+            "table": "ng_styling",
+            "subdirectory": "styling"
+        },
+        {
+            "type": "tabular",
+            "skip": false,
+            "table": "ng_substation_styling",
+            "subdirectory": "substation_styling"
```
Comment (mdhillman) [inline web/uk-base-world/inputs/config/pylons-and-veg/national-grid.json:22]: Subdirectory does not exist, directory is currently named "styling".

**Comment by mdhillman** [comment]
Not sure on the cause, but I cannot spin up the stack to test the visualisation due to the below error from the stack-uploader.

```
10:59:13.009 [main] DEBUG org.apache.hc.client5.http.wire - http-outgoing-3 << "{"ID":"edc6bc1d6712a141751e70335a07f9ed0532ea64617596825ffe0b6f11db9a7b","Running":false,"ExitCode":0,"ProcessConfig":{"tty":false,"entrypoint":"ogr2ogr","arguments":["-f","PostgreSQL","PG:dbname=postgres host=UKBASEWORLD-postgis port=5432 user=postgres password=m0delsrus","/stack_scratch/tmp2518277849561258226","--config","OGR_TRUNCATE","YES","--config","PG_USE_COPY","YES","-t_srs","EPSG:27700","-oo","AUTODETECT_TYPE=YES","-nln","ng_towers","-lco","GEOMETRY_NAME=wkb_geometry","-lco","LAUNDER=NO"],"privileged":false},"OpenStdin":false,"OpenStderr":true,"OpenStdout":true,"CanRemove":false,"ContainerID":"44c52747ecc651856abfc9de2f51f61bc7dd1396e605f9d045b9e18ef8948255","DetachKeys":"","Pid":9169}[\n]"       
10:59:13.009 [main] DEBUG org.apache.hc.client5.http.wire - http-outgoing-3 << "[\r][\n]"
10:59:13.009 [main] DEBUG org.apache.hc.client5.http.headers - http-outgoing-3 << HTTP/1.1 200 OK
10:59:13.009 [main] DEBUG org.apache.hc.client5.http.headers - http-outgoing-3 << Api-Version: 1.43
10:59:13.009 [main] DEBUG org.apache.hc.client5.http.headers - http-outgoing-3 << Content-Type: application/json
10:59:13.009 [main] DEBUG org.apache.hc.client5.http.headers - http-outgoing-3 << Date: Mon, 09 Oct 2023 10:59:13 GMT
10:59:13.009 [main] DEBUG org.apache.hc.client5.http.headers - http-outgoing-3 << Docker-Experimental: false
10:59:13.009 [main] DEBUG org.apache.hc.client5.http.headers - http-outgoing-3 << Ostype: linux
10:59:13.009 [main] DEBUG org.apache.hc.client5.http.headers - http-outgoing-3 << Server: Docker/24.0.5 (linux)
10:59:13.009 [main] DEBUG org.apache.hc.client5.http.headers - http-outgoing-3 << Transfer-Encoding: chunked
10:59:13.009 [main] DEBUG org.apache.hc.client5.http.impl.classic.MainClientExec - ex-0000006E: connection can be kept alive for 3 MINUTES
10:59:13.012 [main] DEBUG org.apache.hc.client5.http.wire - http-outgoing-3 << "0[\r][\n]"
10:59:13.012 [main] DEBUG org.apache.hc.client5.http.wire - http-outgoing-3 << "[\r][\n]"
10:59:13.012 [main] DEBUG org.apache.hc.client5.http.impl.classic.InternalHttpClient - ep-0000006D: releasing valid endpoint
10:59:13.012 [main] DEBUG org.apache.hc.client5.http.impl.io.PoolingHttpClientConnectionManager - ep-0000006D: releasing endpoint
10:59:13.012 [main] DEBUG org.apache.hc.client5.http.impl.io.PoolingHttpClientConnectionManager - ep-0000006D: connection http-outgoing-3 can be kept alive for 3 MINUTES
10:59:13.012 [main] DEBUG org.apache.hc.client5.http.impl.io.PoolingHttpClientConnectionManager - ep-0000006D: connection released [route: {}->unix://localhost:2375][total available: 1; route allocated: 1 of 2147483647; total allocated: 1 of 2147483647]
10:59:13.012 [main] WARN com.cmclinnovations.stack.clients.gdal.GDALClient - Docker exec command returned '0' but wrote the following to stderr:
Warning 6: driver ESRI Shapefile does not support open option AUTODETECT_TYPE

Exception in thread "main" uk.ac.cam.cares.jps.base.exception.JPSRuntimeException: RemoteRDBStoreClient: Failed at closing connection/statement while executing update
        at uk.ac.cam.cares.jps.base.query.RemoteRDBStoreClient.executeUpdate(RemoteRDBStoreClient.java:152)
        at com.cmclinnovations.stack.clients.core.datasets.PostgresDataSubset.runSQLPostProcess(PostgresDataSubset.java:50)
        at com.cmclinnovations.stack.clients.core.datasets.PostgresDataSubset.loadInternal(PostgresDataSubset.java:42)
        at com.cmclinnovations.stack.clients.core.datasets.GeoServerDataSubset.loadInternal(GeoServerDataSubset.java:12)
        at com.cmclinnovations.stack.clients.core.datasets.DataSubset.load(DataSubset.java:54)
        at com.cmclinnovations.stack.clients.core.datasets.DatasetLoader.lambda$loadData$6(DatasetLoader.java:142)
        at java.base/java.util.ArrayList.forEach(Unknown Source)
        at com.cmclinnovations.stack.clients.core.datasets.DatasetLoader.loadData(DatasetLoader.java:142)
        at java.base/java.util.Iterator.forEachRemaining(Unknown Source)
        at com.cmclinnovations.stack.clients.core.datasets.DatasetLoader.uploadInputDatasets(DatasetLoader.java:61)
        at com.cmclinnovations.stack.DataUploader.main(DataUploader.java:12)
Caused by: org.postgresql.util.PSQLException: ERROR: column "voltage" does not exist
  Position: 83
        at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2676)
        at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2366)
        at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:356)
        at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:490)
        at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:408)
        at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:329)
        at org.postgresql.jdbc.PgStatement.executeCachedSql(PgStatement.java:315)
        at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:291)
        at org.postgresql.jdbc.PgStatement.executeUpdate(PgStatement.java:265)
        at uk.ac.cam.cares.jps.base.query.RemoteRDBStoreClient.executeUpdate(RemoteRDBStoreClient.java:150)
        ... 10 more
```

File: web/twa-vis-framework/example-mapbox-vis/webspace/data/singapore/rail-lines.geojson (hunk 1 above)
Comment (Ushcode) [reply web/twa-vis-framework/example-mapbox-vis/webspace/data/singapore/rail-lines.geojson:1]: This was an accidental result of a find and replace of 3838 -> 38383... Reverting!

File: web/uk-base-world/scripts/start.sh (hunk 3 above)
Comment (Ushcode) [reply web/uk-base-world/scripts/start.sh:111]: Forgot to comment this when I'd done so for debugging, thanks

File: web/uk-base-world/scripts/start.sh (hunk 6 above)
Comment (Ushcode) [reply web/uk-base-world/scripts/start.sh:118]: Reverted back to 38383 default

**Comment by Ushcode** [comment]
> National Grid website lists multiple "cables" data sets, please list which one is required (or state that any will do).

Same goes for other READMEs in the data/national_grid directory. 

Individual data links specified for clarity

**Comment by mdhillman** [comment]
New exception in the data uploader:

```
Exception in thread "main" java.lang.RuntimeException: Static GeoServer data '/inputs/data/ukpn/icons' does not exist and could not be loaded.
        at com.cmclinnovations.stack.clients.geoserver.GeoServerClient.loadIcons(GeoServerClient.java:145)
        at com.cmclinnovations.stack.clients.core.datasets.DatasetLoader.loadData(DatasetLoader.java:138)
```

File: web/uk-base-world/scripts/start.sh [hunk 8]
```diff
@@ -98,30 +97,35 @@ then
...
+
+    # Run the stack manager to start a new stack
+    cd "$ROOT/Deploy/stacks/dynamic/stack-manager" || exit
+    echo "Running the stack start up script..."
+    ./stack.sh start UKBASEWORLD 38383
 
     # Copy in the data for upload
     UPLOAD_DATA="$ROOT/Deploy/stacks/dynamic/stack-data-uploader/inputs/data"
+    echo "Copying data..." 
     cp -r "./inputs/data/." "$UPLOAD_DATA/"
```
Comment (mdhillman) [inline web/uk-base-world/scripts/start.sh:110]: The `cd` command here affects the `cp` call later, causing it to copy incorrect directories into the stack uploader. Either need to cd back to $START before line 110, or update the source directory of the `cp` call to `"$START/inputs/data/."`.

**Comment by mdhillman** [comment]
Using updated data directory from Pavilion now generates the following stack uploader error:

```
Exception in thread "main" java.lang.RuntimeException: Couldn't copy '/inputs/data/national_grid/styling' into '/stack_scratch/tmp3954093708225150346' as the source was neither a file nor a directory.
        at com.cmclinnovations.stack.clients.utils.LocalTempDir.copy(LocalTempDir.java:64)
        at com.cmclinnovations.stack.clients.utils.LocalTempDir.copyFrom(LocalTempDir.java:33)
```

However, renaming the `national_grid/substation_styling` directory to the expected `national_grid/styling` then causes a new error:

```
Exception in thread "main" java.lang.RuntimeException: Couldn't copy '/inputs/data/national_grid/substation_styling' into '/stack_scratch/tmp15899268968926965365' as the source was neither a file nor a directory.
        at com.cmclinnovations.stack.clients.utils.LocalTempDir.copy(LocalTempDir.java:64)
        at com.cmclinnovations.stack.clients.utils.LocalTempDir.copyFrom(LocalTempDir.java:33)
```

**Comment by Ushcode** [comment]
This should be ready to merge as long as it is using feature info agent version 3.0.0

**Review by mdhillman — APPROVED** [review]
All comments have been addressed, visualisation has been tested.

**Pull request merged**